pkg/expression/simple_rewriter_test.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/pingcap/tidb/pkg/parser/ast"
+	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/types"
+	"github.com/stretchr/testify/require"
+)
+
+// Original function for comparison
+func findFieldNameOriginal(names types.NameSlice, astCol *ast.ColumnName) (int, error) {
+	dbName, tblName, colName := astCol.Schema, astCol.Table, astCol.Name
+	idx := -1
+	for i, name := range names {
+		if !name.NotExplicitUsable && (dbName.L == "" || dbName.L == name.DBName.L) &&
+			(tblName.L == "" || tblName.L == name.TblName.L) &&
+			(colName.L == name.ColName.L) {
+			if idx != -1 {
+				if names[idx].Redundant || name.Redundant {
+					if !name.Redundant {
+						idx = i
+					}
+					continue
+				}
+				return -1, errNonUniq.GenWithStackByArgs(astCol.String(), "field list")
+			}
+			idx = i
+		}
+	}
+	return idx, nil
+}
+
+func generateTestData(size int) (types.NameSlice, *ast.ColumnName) {
+	names := make(types.NameSlice, size)
+	for i := 0; i < size; i++ {
+		names[i] = &types.FieldName{
+			DBName:  model.NewCIStr("db"),
+			TblName: model.NewCIStr("tbl"),
+			ColName: model.NewCIStr("col" + string(rune('A'+i))),
+		}
+	}
+	astCol := &ast.ColumnName{
+		Schema: model.NewCIStr("db"),
+		Table:  model.NewCIStr("tbl"),
+		Name:   model.NewCIStr("colZ"),
+	}
+	return names, astCol
+}
+
+func TestFindFieldName(t *testing.T) {
+	tests := []struct {
+		name     string
+		names    types.NameSlice
+		astCol   *ast.ColumnName
+		expected int
+		err      error
+	}{
+		{
+			name: "Simple match",
+			names: types.NameSlice{
+				{DBName: model.NewCIStr("db"), TblName: model.NewCIStr("tbl"), ColName: model.NewCIStr("col")},
+			},
+			astCol:   &ast.ColumnName{Schema: model.NewCIStr("db"), Table: model.NewCIStr("tbl"), Name: model.NewCIStr("col")},
+			expected: 0,
+		},
+		{
+			name: "Match with empty schema and table",
```

Comment:
Do we need cases for combinations of schema emptiness x table emptiness?

Reply:
Nice point! Added.